# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

## imports for Python, Pandas

In [81]:
import json
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [3]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [4]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [5]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population info.governor    state shortname
0        Dade       12345    Rick Scott  Florida        FL
1     Broward       40000    Rick Scott  Florida        FL
2  Palm Beach       60000    Rick Scott  Florida        FL
3      Summit        1234   John Kasich     Ohio        OH
4    Cuyahoga        1337   John Kasich     Ohio        OH

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [90]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

{'_id': {'$oid': '52b213b38594d8a2be17c781'},
 'approvalfy': 2015,
 'board_approval_month': 'November',
 'boardapprovaldate': '2013-11-04T00:00:00Z',
 'borrower': 'GOVERNMENT OF TUNISIA',
 'country_namecode': 'Republic of Tunisia!$!TN',
 'countrycode': 'TN',
 'countryname': 'Republic of Tunisia',
 'countryshortname': 'Tunisia',
 'docty': 'Project Information Document,Integrated Safeguards Data Sheet,Integrated Safeguards Data Sheet,Project Information Document,Integrated Safeguards Data Sheet,Project Information Document',
 'envassesmentcategorycode': 'C',
 'grantamt': 4700000,
 'ibrdcommamt': 0,
 'id': 'P144674',
 'idacommamt': 0,
 'impagency': 'MINISTRY OF FINANCE',
 'lendinginstr': 'Specific Investment Loan',
 'lendinginstrtype': 'IN',
 'lendprojectcost': 5700000,
 'majorsector_percent': [{'Name': 'Public Administration, Law, and Justice',
   'Percent': 70},
  {'Name': 'Public Administration, Law, and Justice', 'Percent': 30}],
 'mjsector_namecode': [{'code': 'BX',
   'name': 'Publi

In [7]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia   
1          TN                      Republic of Tunisia          Tunisia   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   

                                      theme_namecode themecode   totalamt  \
0      [{'name': 'Education for all', 'code': '65'}]        65  130000000   
1  [{'name': 'Other economic management', 'code':...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [4]:
#Before we start, let's have a look at the dataset
world_bank = pd.read_json("data/world_bank_projects.json")
world_bank.head(2)

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia   
1          TN                      Republic of Tunisia          Tunisia   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   

                                      theme_namecode themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]        65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

# Question 1: Find the 10 countries with most projects

1. In order to find 10 most-project countries, we need to group data by country names first. Since pd.groupby will not return dataframe directly, “reset_index()" will change pd.Series to pd.DataFrame (you can also use pd.Series.to_frame, but you'll also need to reset index after that);
2. Changing columns to make the results look better;
3. Sort by number of projects and display it from highest to lowest.

In [39]:
#Create dataframe we need
countries = world_bank.groupby('countryshortname').size().reset_index()
#Change column names
countries.columns = ['Country name', 'Number of projects']
#Sort results
countries = countries.sort('Number of projects', ascending = False)
countries.head(10)

/home/yangzhibo5947/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:7: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


Country name  Number of projects
23                China                  19
45            Indonesia                  19
112             Vietnam                  17
44                India                  16
115  Yemen, Republic of                  13
73                Nepal                  12
8            Bangladesh                  12
69              Morocco                  12
70           Mozambique                  11
1                Africa                  11

# Question 2: Find the top 10 major project themes (using column 'mjtheme_namecode')

In [93]:
#Read and take a brief look at json data
#The 'jsons' here just like above json example file, please check above results for detail
#I will not display it here
jsons = json.load((open('data/world_bank_projects.json')))

In [49]:
#Normalizing json data
world_bank_q2 = json_normalize(jsons, 'mjtheme_namecode',['countryname', 'countryshortname'])
#Change the results to dataframe and group by the project name
world_bank_q2_ans = pd.DataFrame({'name':world_bank_q2.name}).groupby(['name']).size()
world_bank_q2_ans

name
                                                122
Economic management                              33
Environment and natural resources management    223
Financial and private sector development        130
Human development                               197
Public sector governance                        184
Rule of law                                      12
Rural development                               202
Social dev/gender/inclusion                     119
Social protection and risk management           158
Trade and integration                            72
Urban development                                47
dtype: int64

We can change the result to dataframe to make it look better

In [52]:
world_bank_q2_df = world_bank_q2_ans.reset_index()
world_bank_q2_df.columns = ['Project name', 'Number of projects']
world_bank_q2_df

Project name  Number of projects
0                                                                122
1                            Economic management                  33
2   Environment and natural resources management                 223
3       Financial and private sector development                 130
4                              Human development                 197
5                       Public sector governance                 184
6                                    Rule of law                  12
7                              Rural development                 202
8                    Social dev/gender/inclusion                 119
9          Social protection and risk management                 158
10                         Trade and integration                  72
11                             Urban development                  47

# Note:
   In above results, we can see there are 122 unnamed projects. By checking the dataset again, it seems we can fill these null values according to its project code. 
   
   That's exactly what Question 3 are asking for.
   
# Question 3 : Sort by number of projects and display it from highest to lowest.

1. In this dataframe, the missing values are represented by empty strings(''), which actually is not missing values in pandas dataframe. So the first thing we need to do is to tansfer empty strings to real null value;
2. Group by project name and show the results.

In [95]:
#Change empty strings to None
world_bank_q2['name'].replace([''], [None], inplace = True)
#Change code column data type from string to int
world_bank_q2['code'] = world_bank_q2['code'].astype(int)
world_bank_q3_df = world_bank_q2.sort('code')
#Fill missing values, ffill means 'backward fill'
world_bank_q3_df.fillna(method = 'bfill', inplace = True)
world_bank_q3_df.head(10)

/home/yangzhibo5947/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:5: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


code                 name                               countryname  \
1114     1  Economic management                      Republic of Honduras   
647      1  Economic management          Lao People's Democratic Republic   
1260     1  Economic management               Hashemite Kingdom of Jordan   
648      1  Economic management          Lao People's Democratic Republic   
1437     1  Economic management                     Republic of Guatemala   
1257     1  Economic management                      Republic of Maldives   
454      1  Economic management  Democratic Republic of Sao Tome and Prin   
453      1  Economic management                        Kingdom of Lesotho   
1057     1  Economic management                    Republic of Azerbaijan   
1024     1  Economic management               United Republic of Tanzania   

                      countryshortname  
1114                          Honduras  
647   Lao People's Democratic Republic  
1260                            Jordan  
648   Lao People's Democratic Republic  
1437                         Guatemala  
1257                          Maldives  
454              Sao Tome and Principe  
453                            Lesotho  
1057                        Azerbaijan  
1024                          Tanzania

In [76]:
#keep only name and code columns
world_bank_q3_ans = world_bank_q3_df.groupby(['code','name']).size().reset_index().sort(0, ascending = False)
world_bank_q3_ans.columns = ['Project code', 'Project name', 'Number of projects']
world_bank_q3_ans

/home/yangzhibo5947/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:2: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  from ipykernel import kernelapp as app


Project code                                  Project name  \
11            11  Environment and natural resources management   
10            10                             Rural development   
8              8                             Human development   
1              2                      Public sector governance   
5              6         Social protection and risk management   
3              4      Financial and private sector development   
7              7                   Social dev/gender/inclusion   
4              5                         Trade and integration   
9              9                             Urban development   
0              1                           Economic management   
2              3                                   Rule of law   
6              7                             Human development   

    Number of projects  
11                 250  
10                 216  
8                  210  
1                  199  
5                  168  
3                  146  
7                  129  
4                   77  
9                   50  
0                   38  
2                   15  
6                    1